// Demonstrates the serialization of an array from server to client
// using gRPC

syntax = "proto3";
package chunkdemo;

message Empty {

}

enum DataType
{
  INTEGER = 0;
  DOUBLE = 1;
}

message RepeatedInts
{
repeated int32 ints = 1;
}

message Chunk
{
  bytes payload = 1;
}

message StreamRequest
{
  int32 chunk_size = 1;
  DataType array_type = 3;
}

message PopulateArrayRequest
{
  int32 array_size = 1;
  DataType array_type = 2;
}

service ChunkDemo{
  // Send a series of chunks containing an array
  rpc DownloadArray(StreamRequest) returns (stream Chunk);
  rpc DownloadArraySlow(StreamRequest) returns (RepeatedInts);
  rpc PopulateArray(PopulateArrayRequest) returns (Empty);

}
